### PR TITLE
Remove functionality for Rack::Utils.key_space_limit

### PR DESCRIPTION
**Pull request by jrochkind**
It was determined that as this limit did not affect nested parameter hashes, it didn't actually prevent an attacker from using more than limited number of bytes for parameter keys, so this limit isn't actually doing anything useful. It is confusing people when it gets in the way of desired large parameter requests.

See notes by and conversation with @jeremyevans at https://github.com/rack/rack/pull/1487

So I decided to take a stab at removing it. 

It ended up being a little bit trickier than simply deprecating `Rack::Utils.key_space_limit=` -- although I have done that. But this parameter had kind of wormed it's way into several different methods and classes. 

I wasn't sure if I should leave every single one for backwards compatibility; in this first draft I do not. 

* changed QueryParser.make_default(key_space_limit, param_depth_limit) to just QueryParser.make_default(param_depth_limit). 

* And changed Rack::QueryParser::Params.new to not take a limit. 

* But left RackQueryParser.new() with it's second positional key_space_limit arg (which just no-ops and issues deprecation message), because it seems somewhat more likely to be used by third party code, and changing a 3-positional-arg signature to 2 by removing the middle one seems especially confusing. 

* Removed QueryParser.key_space_limit reader, but left QueryParser.key_space_limit= writer as a no-op deprecated, because we KNOW the latter is used by callers wanting to increase it!

I can change these choices if you let me know what you'd prefer. Leave ALL of that stuff, with deprecation warnings, like `Rack::QueryParser::Params.new(arg)` would get you a deprecation warning? I can do that!

It's also not clear to me how this fits into Rack release plans (if any). CHANGELOG makes it look like there's an upcoming 3.0 release... but we shouldn't put in deprecations in a 3.0 release... so would that mean this should actually be PR'd to a 2.x patch release? Not sure if I should change my PR base, or what. Didn't put anything in CHANGELOG cause didn't know what was going on with that. 

As usual with software like this, what seemed like it shoudl be a simple change ended up tangled in a few different places/choices.... might be more trouble than it's worth? But I thought I'd at least make a stab at it, feedback welcome.

### CODE REVIEW

**Comment by jrochkind** [comment]
It might be good to get this in before Rack 3, and actually removed in Rack 3?

Happy to take feedback on right approach -- maybe we don't even need to deprecate anything, and can change method signatures here with abandon, for Rack 3?

**Comment by jeremyevans** [comment]
> I wasn't sure if I should leave every single one for backwards compatibility; in this first draft I do not.

I think backwards compatibility should be kept in most of these cases.
 
> * changed QueryParser.make_default(key_space_limit, param_depth_limit) to just QueryParser.make_default(param_depth_limit).

Probably best to do:

```ruby
QueryParser.make_default(_key_space_limit=(deprecated = true; nil), param_depth_limit)
```

And warn if `deprecated`.

> * And changed Rack::QueryParser::Params.new to not take a limit.

We could do:

```ruby
def initialize(limit=(deprecated = true; nil))
```

However, this is one case where we may want to drop backwards compatibility as this is called in a hot path and I don't think keeping backwards compatibility is worth the loss in performance.

> * But left RackQueryParser.new() with it's second positional key_space_limit arg (which just no-ops and issues deprecation message), because it seems somewhat more likely to be used by third party code, and changing a 3-positional-arg signature to 2 by removing the middle one seems especially confusing.

```ruby
def initialize(params_class, _key_space_limit=(deprecated = true; nil), param_depth_limit)
```

> * Removed QueryParser.key_space_limit reader, but left QueryParser.key_space_limit= writer as a no-op deprecated, because we KNOW the latter is used by callers wanting to increase it!

I would leave the reader and also have it issue a warning.

> I can change these choices if you let me know what you'd prefer. Leave ALL of that stuff, with deprecation warnings, like `Rack::QueryParser::Params.new(arg)` would get you a deprecation warning? I can do that!

The above are just my opinions, it may be a good idea to get feedback from other committers before making any changes to the PR.  One possible alternative is making the changes you want, and adding a separate file you can require that will prepend modules to the appropriate classes to make the API backwards compatible and result in deprecation warnings instead of errors.

> It's also not clear to me how this fits into Rack release plans (if any). CHANGELOG makes it look like there's an upcoming 3.0 release... but we shouldn't put in deprecations in a 3.0 release... so would that mean this should actually be PR'd to a 2.x patch release? Not sure if I should change my PR base, or what. Didn't put anything in CHANGELOG cause didn't know what was going on with that.
> 
> It might be good to get this in before Rack 3, and actually removed in Rack 3?

That seems unlikely.  My understanding is the next release will be 3.0.0, unless there is a security issue that causes the release of a point version.

> Happy to take feedback on right approach -- maybe we don't even need to deprecate anything, and can change method signatures here with abandon, for Rack 3?

I think we should deprecate in Rack 3 and remove in Rack 4.

**Comment by jrochkind** [comment]
Thank you for feedback @jeremyevans , I appreciate it! 

I will try to wait for feedback from other committers before making changes; if I don't get any in a few weeks, may go ahead and implement yours. 

I do not understand this code:

    def initialize(limit=(deprecated = true; nil))

OK.... it is setting the default value of `limit` to nil, while as a side effect setting local variable `deprecated` to true? So if local variable `deprecate` is set to true, we know that the default value was used.... is this then meant to control whether deprecation warning is printed? So in this case you'd print the warning only if `deprecated` is _not_ true.... so perhaps variable should be named differently, this is getting confusing?

Am I following that right? That is considered preferable to just issuing the warning if `!limit.nil?`?  I'm not following why?

Note also that this method of side effects in default args is only avaialable on the _last_ positional args.  So we can't actually do it on `self.make_default(key_space_limit, param_depth_limit)`, right? There's no way to supply a `param_depth_limit` while allowing the default value of `key_space_limit` to be used, since you can't "skip" a positional arg when calling. So we'd really have to use the "warn if non-nil value is present" approach there anyway, right?

**Comment by jeremyevans** [comment]
> Thank you for feedback @jeremyevans , I appreciate it!
> 
> I will try to wait for feedback from other committers before making changes; if I don't get any in a few weeks, may go ahead and implement yours.
> 
> I do not understand this code:
> 
> ```
> def initialize(limit=(deprecated = true; nil))
> ```
> 
> OK.... it is setting the default value of `limit` to nil, while as a side effect setting local variable `deprecated` to true? So if local variable `deprecate` is set to true, we know that the default value was used.... is this then meant to control whether deprecation warning is printed? So in this case you'd print the warning only if `deprecated` is _not_ true.... so perhaps variable should be named differently, this is getting confusing?

Correct, it should be `def initialize(limit=(not_deprecated = true; nil))`, and warn unless `not_deprecated`

> Am I following that right? That is considered preferable to just issuing the warning if `!limit.nil?`? I'm not following why?

`limit.nil?` doesn't catch cases where `nil` is passed explicitly as an argument.  If you want to remove a method parameter in a backwards compatible manner, you have to issue deprecation warnings based on whether an argument was passed.  You cannot just check the argument matches the default value.

> Note also that this method of side effects in default args is only avaialable on the _last_ positional args. So we can't actually do it on `self.make_default(key_space_limit, param_depth_limit)`, right? There's no way to supply a `param_depth_limit` while allowing the default value of `key_space_limit` to be used, since you can't "skip" a positional arg when calling. So we'd really have to use the "warn if non-nil value is present" approach there anyway, right?

With:

```ruby
QueryParser.make_default(_key_space_limit=(not_deprecated = true; nil), param_depth_limit)
```

If you pass only a single argument, it sets `param_depth_limit`, and uses the default value for `_key_space_limit`, which, will set `not_deprecated`.  You cannot skip a mandatory positional argument, but you can skip an optional positional argument when setting just a post argument.  This has been true since Ruby 1.9.  Example:

```ruby
Object.make_default(_key_space_limit=(not_deprecated = true; nil), param_depth_limit)
  [_key_space_limit, not_deprecated, param_depth_limit]
end

 Object.make_default(1)
# => [nil, true, 1]

Object.make_default(nil, 1)
# => [nil, nil, 1]

Object.make_default(2, 1)
# => [2, nil, 1]
```

**Comment by jrochkind** [comment]
> but you can skip an optional positional argument when setting just a post argument. This has been true since Ruby 1.9. Example:

Wow, I had no idea, thanks!

To me, I'm not sure this makes the code less confusing rather than more -- especially for callers, who have to figure out they can call with one param and what it will do if they do so. 

But whatever you and other committers say, I will implement! So far it's this, ok! And thanks for teaching me a new trick!

**Review by ioquatix — APPROVED** [review]

**Comment by ioquatix** [comment]
@jrochkind How much cleaning up is needed after this is merged? i.e. what is the process of deprecation?

**Comment by jrochkind** [comment]
Oh gosh, I don't even remember this anymore from 7 months ago. 

Looking at the conversation, it looks like this PR does not maintain as much strict backwards compatibility as @jeremyevans wanted. 

I can go back and do what he suggested and put more backwards compat back in, if there is a good chance it will actually get merged. I just didn't want to do the work and have it sit here, right?

I will include deprecation messages as necessary -- the consensus is the deprecated functionality didn't do anything useful, which is why it's being removed. So I don't know what other additional cleanup would be necessary?

But ok, if there's actually some decent chance it will get merged, I'll try to do that additional work this week, thanks!

**Comment by ioquatix** [comment]
I personally support merging this, but we need to:

1. Rebase.
2. Address feedback.

@jeremyevans are you on board with this?

**Comment by jeremyevans** [comment]
> I personally support merging this, but we need to:
> 
> 1. Rebase.
> 2. Address feedback.
> 3. A little bit of documentation regarding longer term deprecation process/followup changes.
> 
> @jeremyevans are you on board with this?

Yes, I'm on board with this approach.

**Comment by jrochkind** [comment]
Thank you! I will work on this presently. 

> A little bit of documentation regarding longer term deprecation process/followup changes.

If you let me know where such docs go, I can take a stab at it.

**Comment by jrochkind** [comment]
@jeremyevans You suggest that the `QueryParser.key_space_limit` reader should remain but with a deprecation warning. 

However, this PR actually *removes* the value that it would return, because it would have no effect. 

Should I keep new state as set by the (deprecated) writer, that will do nothing except be there to be returned by the (deprecated) reader? 

I guess I will try to do that. Not sure where to put that state though. I'll try to figure it out, or please advise welcome.

**Comment by jeremyevans** [comment]
> Should I keep new state as set by the (deprecated) writer, that will do nothing except be there to be returned by the (deprecated) reader?

I don't think that is necessary.  Maybe have it return the previous default value (65536)?  Then there is no need to keep state.

**Comment by jeremyevans** [comment]
Looks good to me.  The test failures are due to ubuntu packages not installing, which is out of our control.

**Comment by jrochkind** [comment]
Thank you @jeremyevans !

**Pull request merged**
